### PR TITLE
Add naive server

### PR DESCRIPTION
**Pull request by lbouchard-adgear**
Would like to have your opinion on this naive server implementation to improve my style.  

Structure and design comments are also welcomed and could be a nice complement to the following PR:
https://github.com/adgear/barker-go/pull/2

I am not sure I will merge to master yet as I already have more advanced code for the server. It might be useful to have a bit of code in place though, even if it is naive.

### CODE REVIEW

File: client/go.mod [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module client
```
Comment (pdeschen) [inline client/go.mod:1]: Do we really need a client? maybe a `cli` later down the road but for now I feel we could just create unit tests?

File: barker/go.mod [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module barker
```
Comment (pdeschen) [inline barker/go.mod:1]:
```bash
├── barker
│   ├── go.mod
│   └── main.go
├── client
│   ├── go.mod
│   └── main.go
└── server
    ├── go.mod
    ├── handler
    │   ├── dummyhandler
    │   │   └── dummyhandler.go
    │   └── handler.go
    └── server
        └── server.go
```

I'm no go expert and a bit rusty with this but I feel we could actually remove `barker` subfolder, move the `main.go` under `server` and move content of `server` up to the root and then transform `client` into an actual test suite within `test` folder.

As it stands, we have `server\server\server.go`

File: client/go.mod (hunk 1 above)
Comment (lbouchard-adgear) [reply client/go.mod:1]: You must really love coding to be here 😄

Comment (lbouchard-adgear) [reply client/go.mod:1]: Personally, I like to have a simple way to run the application and test it end-to-end, i.e. not just with unit tests. 
But well, if you guys prefer to have only a unit test or have both a unit test and a client,  I don't mind.

File: server/server/server.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  address string
+  socket net.Listener
+  maxBufferSize int
+  requestHandler handler.RequestHandler
+}
+
+
+////////////
+// Public //
+////////////
```
Comment (pdeschen) [inline server/server/server.go:23]: I believe you don't need those. Capitalized vs non capitalized exports gives you that info.

File: barker/go.mod (hunk 2 above)
Comment (lbouchard-adgear) [reply barker/go.mod:1]: Good point for server/server ... I think I could probably bring back server.go directory under `server` instead `server/server`.  Will wait to have a bit more feedback for the main location.

For the unit tests, if I am not mistaking, I think that it is a common pattern to put the test file within the same directory as the module file. For example, there would be server_test.go just next to server.go.   I think that the global test folder is typically kept for other types of tests (ex: integration tests, performance tests, etc.)

Comment (lbouchard-adgear) [reply barker/go.mod:1]: But well, I could probably write a basic test mimicking what is done in the client.

Comment (pdeschen) [reply barker/go.mod:1]: Yes. Test leaves alongside code within the package. Sorry I wasn’t clear enough. 

Yep a small test boilerplate would be nice

File: client/go.mod (hunk 1 above)
Comment (pdeschen) [reply client/go.mod:1]: Yes I miss coding......lol

Don’t get me wrong: I think a small cli tool would make sense at least to showcase features within the server as part of being demoable. But it’s more a cli than an actual client lib. And that could be later down the once we actually implement feature and then have to demo it. This way could/should provision time for it. 

Anyways just a thought. No big deal

If you can’t trust unit test then how can you trust the code? :-)

File: server/server/server.go (hunk 3 above)
Comment (lbouchard-adgear) [reply server/server/server.go]: I remove the "public" block.  I think that it is indeed not needed.

I kept the block for the private section though.  I like, personally, to have a big visual marker to see when private functions start. You guys tell me what you prefer.

File: client/go.mod (hunk 1 above)
Comment (lbouchard-adgear) [reply client/go.mod:1]: I'm picky :) . I think that unit tests are great to test isolated piece of code.  Sometimes, though,  they fail to capture integration problems.

Comment (lbouchard-adgear) [reply client/go.mod]: Just to make sure that there is no misunderstanding here, the "client" here is not meant to be a client library that is imported in some other golang project.  It is an executable that is meant to be run with `go run client`.

Comment (lbouchard-adgear) [reply client/go.mod]: Ended up moving the client code to a global test folder. This way it is clear that it is a testing tool for now.
Added a unit test `server_test.go`

File: barker/go.mod (hunk 2 above)
Comment (lbouchard-adgear) [reply barker/go.mod:1]: Still curious to know what other think about package structure.  My intention with keeping the server package separated from the rest was to emphasize that I want it decoupled from the rest.  I'd like that we can reuse the same server logic for different types of model / optimizers.

Comment (lbouchard-adgear) [reply barker/go.mod:1]: As mentioned in another thread, wrote a unit test for server and move the client code to the global test folder.

File: server/server/server.go (hunk 3 above)
Comment (psyomn) [reply server/server/server.go]: Capitalization denotes the public methods. It sounds weird but works pretty well. 

As for grouping of the interface, up to you. If it were up to me, I'd keep most of my public facing methods in a file, which would define the proper interface for the consumer, and move all the gnarly stuff in some private file. 

I didn't read your code, but I'm just giving my thoughts on the matter (so you might be doing this already)

**Comment by lbouchard-adgear** [comment]
Thank you all for your time and useful comments :) .

**Pull request closed without merging**